csrc/executor.h

```diff
@@ -247,13 +275,49 @@ class FusionExecutor : public NonCopyable {
         compiled_kernel_->cubin, "-fun 1 -c");
   }
 
-  std::string getCanonicalKernelName() const {
-    return kernelNamespace() + "::" + kernelName();
+  static void setGlobalFusionCount(int64_t new_fusion_count) {
+    global_fusion_count_.store(new_fusion_count);
+  }
+
+  static int64_t getGlobalFusionCount() {
+    return global_fusion_count_.load();
+  }
+
+  void createKernelId(
+      ScheduleHeuristic heuristic = ScheduleHeuristic::None,
+      int64_t fusion_id = 0,
+      int64_t concrete_id = 0,
+      int64_t runtime_id = 0,
+      int64_t group_id = 0) {
+    NVF_ERROR(fusion_id > -1, "Invalid fusion_id.");
+    NVF_ERROR(concrete_id > -1, "Invalid concrete_id.");
+    NVF_ERROR(runtime_id > -1, "Invalid runtime_id.");
+    NVF_ERROR(group_id > -1, "Invalid group_id");
+
+    heuristic_ = heuristic;
+    fusion_id_ = fusion_id;
+    concrete_id_ = concrete_id;
+    runtime_id_ = runtime_id;
+    group_id_ = group_id;
+    ++global_fusion_count_;
+
+    std::stringstream ss;
+    if (isOptionEnabled(EnableOption::StaticFusionCount)) {
+      ss << global_fusion_count_.load();
+    } else {
+      ss << toString(heuristic_);
+      ss << "_f" << fusion_id_;
+      ss << "_c" << concrete_id_;
+      ss << "_r" << runtime_id_;
+      ss << "_g" << group_id_;
+    }
+    kernel_id_ = ss.str();
   }
 
   std::string kernelName() const {
+    NVF_ERROR(!kernel_id_.empty(), "Invalid kernel name for fusion executor.");
     std::stringstream ss;
-    ss << "kernel" << fusion_id_;
+    ss << "nvfuser_" << kernel_id_;
```

Comment:
interesting that we are not giving the anonymous namespace here, yet nvrtc is correctly finding the name?! :confused:

Reply:
This part looks a bit scary to me. What's the reason to have the anonymous namespace in the first place? I have limited understanding on our compiler toolchain and I'm surprised that function in an anonymous namespace is accessible at all.

Should we just remove the namespace?! Is there any concern we have about leaking thing to global namespace in nvrtc? Since we don't have any headers here, I'm guessing this is low risk?!?! cc'ing folks who actually knows what we have in our pasted headers. @naoyam @jacobhinkle @zasdfgbnm

Reply:
I thought the point of the namespace was purely to make sure that none of the runtime code collides with anything outside of the namespace and it should be a single translation unit as you are giving all of the code to NVRTC in one file (string).  Therefore, the name of the namespace is unneeded.

Reply:
Good point. ^

This reads about right as well here: https://docs.nvidia.com/cuda/nvrtc/index.html#accessing-lowered-names
>  During compilation, NVRTC will parse the name expression string as a C++ constant expression at the end of the user program. The constant expression must provide the address of the __global__ function or __device__/__constant__ variable.

So it's just an extra piece attached after the user program and is compiled together as a constant string. Sounds good to me then.